File: coding examples/Data Envelopment Analysis based Efficiency/MATLAB/dea.m
I = [150 70 5;
   60 20 3;
   800 140 20;
   30 15 1];

O = [15 225;
    5.4 70;
    56 1300;
    2.1 40];

[m n1]=size(I);
[m n2]=size(O);

% x is [v1;v2,...,vn1;u1;u2...,un2]
f=[zeros(m,n1) -O];
lb=zeros(n1+n2,1);

x=zeros(m,n1+n2);
A=[zeros(m,n1) O] - [I zeros(m,n2)];
b=zeros(m,1);

for i=1:m
    Aeq=[I(i,:) zeros(1,n2)];
    beq=1;
    z = linprog(f(i,:),A,b,Aeq,beq,lb);
    x(i,:)=z';
end
result=x.*[I O];
sum(result(:,1:3),2)
sum(result(:,4:5),2)

File: machine learning/Data Envelopment Analysis based Efficiency/R/dea.R
library(linprog)

I = matrix(c(150,60,800,30,70,20,140,15,5,3,20,1), nrow = 4, ncol = 3)
O = matrix(c(15,5.4,56,2.1,225,70,1300,40),nrow =4, ncol = 2)
a = dim(I)
m = a[1]
n1 = a[2]

b = dim(O)
n2 = b[2]
f = cbind(matrix(0,m,n1),O)
lb = matrix(0,n1+n2,1)
x = matrix(0,m,n1+n2)

for (i in 1:m)
{
  Aeq = c(I[i,], matrix(0,1,n2))
  beq = 1
  A = rbind(Aeq,cbind(matrix(0,m,n1), O) - cbind(I, matrix(0,m,n2)))
  b = rbind(1,matrix(0,m,1))
  z = solveLP(f[i,],b,A,maximum=TRUE,const.dir = c("==","<=","<=","<=","<="),lpSolve=TRUE)
  x[i,]=z$solution
}

result = x*cbind(I,O)
sum(result[,1:3])
rowSums(result[,4:5])
